File: other/past201912-open/L.py
import sys
from itertools import combinations
from operator import itemgetter

sys.setrecursionlimit(10 ** 7)
rl = sys.stdin.readline


class UnionFind:
    def __init__(self, n: int):
        self.n = n
        self.parents = [-1] * n
    
    def find(self, x: int):
        if self.parents[x] < 0:
            return x
        else:
            self.parents[x] = self.find(self.parents[x])
            return self.parents[x]
    
    def union(self, x: int, y: int):
        x = self.find(x)
        y = self.find(y)
        
        if x == y:
            return
        if self.parents[y] < self.parents[x]:
            x, y = y, x
        
        self.parents[x] += self.parents[y]
        self.parents[y] = x
    
    def size(self, x: int):
        return -self.parents[self.find(x)]
    
    def same(self, x: int, y: int):
        return self.find(x) == self.find(y)
    
    def members(self, x: int):
        root = self.find(x)
        return [i for i in range(self.n) if self.find(i) == root]
    
    def roots(self):
        return [i for i, x in enumerate(self.parents) if x < 0]
    
    def group_count(self):
        return len(self.roots())
    
    def all_group_members(self):
        return {r: self.members(r) for r in self.roots()}
    
    def __str__(self):
        return '\n'.join('{}: {}'.format(r, self.members(r)) for r in self.roots())


def solve():
    N, M = map(int, rl().split())
    x, y, c = [0] * N, [0] * N, [0] * N
    for i in range(N):
        x[i], y[i], c[i] = map(int, rl().split())
    X, Y, C = [0] * M, [0] * M, [0] * M
    for i in range(M):
        X[i], Y[i], C[i] = map(int, rl().split())
    
    big_tower_edges = []
    for i, j in combinations(range(N), 2):
        cost = ((x[i] - x[j]) ** 2 + (y[i] - y[j]) ** 2) ** 0.5
        if c[i] != c[j]:
            cost *= 10
        big_tower_edges.append((cost, i, j))
    
    ans = 10 ** 18
    for s in range(1 << M):
        small_towers = []
        for i in range(M):
            if s >> i & 1:
                small_towers.append(i)
        add_edges = []
        for i in small_towers:
            for j in range(N):
                cost = ((X[i] - x[j]) ** 2 + (Y[i] - y[j]) ** 2) ** 0.5
                if C[i] != c[j]:
                    cost *= 10
                add_edges.append((cost, i + N, j))
        for i, j in combinations(small_towers, 2):
            cost = ((X[i] - X[j]) ** 2 + (Y[i] - Y[j]) ** 2) ** 0.5
            if C[i] != C[j]:
                cost *= 10
            add_edges.append((cost, i + N, j + N))
        current_edges = big_tower_edges + add_edges
        current_edges.sort(key=itemgetter(0))
        
        uf = UnionFind(N + M)
        tmp = 0
        for cost, i, j in current_edges:
            if not uf.same(i, j):
                tmp += cost
                uf.union(i, j)
        ans = min(ans, tmp)
    print(ans)


if __name__ == '__main__':
    solve()

File: other/past202104-open/L.nim
import
  algorithm, bitops, deques, heapqueue, math, macros, sets, sequtils,
  strformat, strutils, sugar, tables

proc transLastStmt(n, res, bracketExpr: NimNode): (NimNode, NimNode, NimNode) =
  # Looks for the last statement of the last statement, etc...
  case n.kind
  of nnkIfExpr, nnkIfStmt, nnkTryStmt, nnkCaseStmt:
    result[0] = copyNimTree(n)
    result[1] = copyNimTree(n)
    result[2] = copyNimTree(n)
    for i in ord(n.kind == nnkCaseStmt)..<n.len:
      (result[0][i], result[1][^1], result[2][^1]) = transLastStmt(n[i], res, bracketExpr)
  of nnkStmtList, nnkStmtListExpr, nnkBlockStmt, nnkBlockExpr, nnkWhileStmt,
      nnkForStmt, nnkElifBranch, nnkElse, nnkElifExpr, nnkOfBranch, nnkExceptBranch:
    result[0] = copyNimTree(n)
    result[1] = copyNimTree(n)
    result[2] = copyNimTree(n)
    if n.len >= 1:
      (result[0][^1], result[1][^1], result[2][^1]) = transLastStmt(n[^1], res, bracketExpr)
  of nnkTableConstr:
    result[1] = n[0][0]
    result[2] = n[0][1]
    if bracketExpr.len == 1:
      bracketExpr.add([newCall(bindSym"typeof", newEmptyNode()), newCall(
          bindSym"typeof", newEmptyNode())])
    template adder(res, k, v) = res[k] = v
    result[0] = getAst(adder(res, n[0][0], n[0][1]))
  of nnkCurly:
    result[2] = n[0]
    if bracketExpr.len == 1:
      bracketExpr.add(newCall(bindSym"typeof", newEmptyNode()))
    template adder(res, v) = res.incl(v)
    result[0] = getAst(adder(res, n[0]))
  else:
    result[2] = n
    if bracketExpr.len == 1:
      bracketExpr.add(newCall(bindSym"typeof", newEmptyNode()))
    template adder(res, v) = res.add(v)
    result[0] = getAst(adder(res, n))

macro collect*(init, body: untyped): untyped =
  runnableExamples:
    import sets, tables
    let data = @["bird", "word"]
    ## seq:
    let k = collect(newSeq):
      for i, d in data.pairs:
        if i mod 2 == 0: d
    assert k == @["bird"]
    ## seq with initialSize:
    let x = collect(newSeqOfCap(4)):
      for i, d in data.pairs:
        if i mod 2 == 0: d
    assert x == @["bird"]
    ## HashSet:
    let y = initHashSet.collect:
      for d in data.items: {d}
    assert y == data.toHashSet
    ## Table:
    let z = collect(initTable(2)):
      for i, d in data.pairs: {i: d}
    assert z == {0: "bird", 1: "word"}.toTable
  
  let res = genSym(nskVar, "collectResult")
  expectKind init, {nnkCall, nnkIdent, nnkSym}
  let bracketExpr = newTree(nnkBracketExpr,
    if init.kind == nnkCall: init[0] else: init)
  let (resBody, keyType, valueType) = transLastStmt(body, res, bracketExpr)
  if bracketExpr.len == 3:
    bracketExpr[1][1] = keyType
    bracketExpr[2][1] = valueType
  else:
    bracketExpr[1][1] = valueType
  let call = newTree(nnkCall, bracketExpr)
  if init.kind == nnkCall:
    for i in 1 ..< init.len:
      call.add init[i]
  result = newTree(nnkStmtListExpr, newVarStmt(res, call), resBody, res)

proc input*(): string {.inline.} = stdin.readLine
proc inputs*(): seq[string] {.inline.} = stdin.readLine.split
proc inputInt*(): int {.inline.} = stdin.readLine.parseInt
proc inputInts*(): seq[int] {.inline.} = stdin.readLine.split.map(parseInt)
proc chmax*[T: SomeNumber](n: var T, m: T) {.inline.} = n = max(n, m)
proc chmin*[T: SomeNumber](n: var T, m: T) {.inline.} = n = min(n, m)
proc `%=`*[T: SomeInteger](n: var T, m: T) {.inline.} = n = floorMod(n, m)

when not declared ATCODER_UNIONFIND_HPP:
  const ATCODER_UNIONFIND_HPP* = 1

  type
    UnionFind* = ref object
      n: Positive
      parents: seq[int]

  proc initUnionFind*(n: Positive): UnionFind =
    return UnionFind(n: n, parents: newSeqWith(n, -1))

  proc find*(self: var UnionFind, x: Natural): Natural =
    if self.parents[x] < 0:
      return x
    else:
      self.parents[x] = self.find(self.parents[x])
      return self.parents[x]

  proc union*(self: var UnionFind, x, y: Natural) =
    var
      root_x = self.find(x)
      root_y = self.find(y)
    
    if root_x == root_y:
      return
    if self.parents[root_y] < self.parents[root_x]:
      (root_x, root_y) = (root_y, root_x)
    self.parents[root_x] += self.parents[root_y]
    self.parents[root_y] = root_x

  proc size*(self: var UnionFind, x: Natural): Positive =
    return -self.parents[self.find(x)]

  proc same*(self: var UnionFind, x, y: Natural): bool =
    return self.find(x) == self.find(y)

  proc members*(self: var UnionFind, x: Natural): seq[int] =
    let root = self.find(x)
    return toSeq(0..<int(self.n)).filterIt(self.find(it) == root)

  proc roots*(self: var UnionFind): seq[int] =
    return toSeq(0..<int(self.n)).filterIt(self.parents[it] < 0)

  proc groupCount*(self: var UnionFind): Positive =
    return self.roots.len

when isMainModule:
  var
    N, M: int
    xi, yi, ri: float
  (N, M) = inputInts()
  let
    P = collect(newSeq):
      for _ in 0..<N: (xi, yi) = inputs().map(parseFloat); (xi, yi)
    C = collect(newSeq):
      for _ in 0..<M: (xi, yi, ri) = inputs().map(parseFloat); (xi, yi, ri)
  
  proc weight(x1, y1, r1, x2, y2, r2: float): float =
    let d = hypot(x1 - x2, y1 - y2)
    if d > r1 + r2:
      return d - r1 - r2
    elif d <= r1 + r2 and abs(r1 - r2) > d:
      return abs(r1 - r2) - d
    else:
      return 0.0
  
  var
    edges = newSeq[(int, int, float)]()
    x1, y1, r1, x2, y2, r2: float
  for i in 0..<N - 1:
    (x1, y1) = P[i]
    for j in i + 1..<N:
      (x2, y2) = P[j]
      edges.add((i + M, j + M, hypot(x1 - x2, y1 - y2)))
  for i in 0..<N:
    (x1, y1) = P[i]
    for j in 0..<M:
      (x2, y2, r2) = C[j]
      edges.add((i + M, j, weight(x1, y1, 0.0, x2, y2, r2)))
  for i in 0..<M - 1:
    (x1, y1, r1) = C[i]
    for j in i + 1..<M:
      (x2, y2, r2) = C[j]
      edges.add((i, j, weight(x1, y1, r1, x2, y2, r2)))
  edges.sort((a, b) => (if a[2] > b[2]: 1 else: -1))

  var ans = 1e18
  for S in 0..<1 shl M:
    var node = toSeq(M..<M + N).toHashSet
    for i in 0..<M:
      if bitand(S shr i, 1) == 1:
        node.incl(i)
    var
      uf = initUnionFind(N + M)
      cost: float
    for (u, v, c) in edges:
      if u notin node or v notin node:
        continue
      if uf.same(u, v):
        continue
      uf.union(u, v)
      cost += c
      if uf.size(u) == node.len:
        break
    ans.chmin(cost)
  echo ans
